// dbdocs/src/utils/constants.js
const PROJECT_GENERAL_ACCESS_TYPE = {
  public: 'public',
  protected: 'protected',
  restricted: 'restricted',
};

const PROJECT_SHARING_TEXT = {
  public: 'Public',
  protected: 'Password-protected',
  restricted: 'Private',
};

const FLAG_HELP_GROUP = {
  sharing: 'sharing',
};

const LOGIN_METHODS = {
  EMAIL: 'Email',
  GITHUB_GOOGLE: 'GitHub/Google',
};

const SUPPORTED_DATABASE_CONNECTORS = ['postgres', 'mysql', 'mssql', 'oracle', 'snowflake', 'bigquery'];

/**
 * Match window path
 * Eg:
 * - C:\Documents\code\credential.json
 * - C:/Documents/code/credential.json
 */
const WINDOW_FILE_PATH_REGEX = /^[a-zA-Z]:[\\/](?:[^<>:"/\\|?*\n\r]+[\\/])*[^<>:"/\\|?*\n\r]*$/;

/**
 * Match unix system path
 * Eg:
 * - /Users/dev/code/credential.json
 * - ~/code/credential.json
 * - ./code/credential.json
 */
const UNIX_FILE_PATH_REGEX = /^(\/|\.\/|~\/|\.\.\/)([^<>:"|?*\n\r]*\/?)*[^<>:"|?*\n\r]*$/;

const HOST_CONFIG_STATUS = {
  active: 'active',
  inactive: 'inactive',
};

const LOGIN_STRATEGIES = {
  legacy: 'legacy',
  portal: 'portal',
};

/**
 * Portal authentication error codes
 * Used for OAuth 2.0 flow errors
 */
const PORTAL_ERROR_CODES = {
  loginTimeout: 'LOGIN_TIMEOUT',
  loginCancelled: 'LOGIN_CANCELLED',
  stateMismatch: 'STATE_MISMATCH',
  invalidCallback: 'INVALID_CALLBACK',
  exchangeFailed: 'EXCHANGE_FAILED',
  loopbackServerError: 'LOOPBACK_SERVER_ERROR',
  unknownError: 'UNKNOWN_ERROR',
};

const PORTAL_LOGIN_CONFIGS = {
  clientId: 'dbdocs-cli',
  loginTimeoutMs: 5 * 60 * 1000, // 5 minutes
  loopbackServer: {
    // It's recommended to use 127.0.0.1 instead of localhost since the localhost can be overriden on the host machine
    // See https://www.rfc-editor.org/rfc/rfc8252#section-8.3
    address: '127.0.0.1',
    // For node http server, port 0 means the OS will assign a random one
    autoAssignedPort: 0,
  },
};

module.exports = {
  PROJECT_GENERAL_ACCESS_TYPE,
  PROJECT_SHARING_TEXT,
  FLAG_HELP_GROUP,
  LOGIN_METHODS,
  SUPPORTED_DATABASE_CONNECTORS,
  WINDOW_FILE_PATH_REGEX,
  UNIX_FILE_PATH_REGEX,
  HOST_CONFIG_STATUS,
  LOGIN_STRATEGIES,
  PORTAL_ERROR_CODES,
  PORTAL_LOGIN_CONFIGS,
};
