// dbdocs/src/vars.js
const { LOGIN_STRATEGIES } = require("./utils/constants");

/* eslint-disable class-methods-use-this */
class Vars {
  get host () {
    return this.envHost || 'dbdocs.io';
  }

  get apiHost () {
    return this.envApiHost || `api.${this.host}`;
  }

  get hostUrl () {
    return this.host.startsWith('http') ? this.host : `https://${this.host}`;
  }

  get apiUrl () {
    return this.apiHost.startsWith('http') ? this.apiHost : `https://${this.apiHost}`;
  }

  get envHost () {
    return process.env.DBDOCS_HOST;
  }

  get envApiHost () {
    return process.env.DBDOCS_API_HOST;
  }

  get loginStrategy () {
    const envLoginStrategy = process.env.DBDOCS_LOGIN_STRATEGY || '';
    return Object.values(LOGIN_STRATEGIES).includes(envLoginStrategy) ? envLoginStrategy : LOGIN_STRATEGIES.portal;
  }

  get loginUrl () {
    return process.env.DBDOCS_LOGIN_URL || 'https://dbdiagram.io';
  }

  get loginApiUrl () {
    return process.env.DBDOCS_LOGIN_API_URL || this.apiUrl;
  }
}

module.exports.Vars = Vars;
module.exports.vars = new Vars();
/* eslint-enable class-methods-use-this */
