// dbdocs/src/services/auth/portal/index.js
const { randomBytes, createHash } = require('crypto');
const axios = require('axios');
const { startLoopbackServer } = require('./server');
const { PORTAL_ERROR_CODES, PORTAL_LOGIN_CONFIGS } = require('../../../utils/constants');
const { PortalAuthenticationError } = require('../../../errors/common-errors');

const STATE_BYTES_LENGTH = 32;
const CODE_VERIFIER_BYTES_LENGTH = 32;

const generateLoginState = () => randomBytes(STATE_BYTES_LENGTH).toString('base64url');

const generateCodeVerifier = () => randomBytes(CODE_VERIFIER_BYTES_LENGTH).toString('base64url');

/**
 *
 * @param {string} codeVerifier
 * @returns {string}
 */
const generateCodeChallenge = (codeVerifier) => {
  const hash = createHash('sha256');
  hash.update(codeVerifier);

  return hash.digest().toString('base64url');
};

/**
 * Build portal login URL with OAuth parameters
 * @param {{ baseUrl: string, redirectUri: string, codeChallenge: string, state: string, clientId: string }} params
 * @returns {URL}
 */
const generateLoginUrl = (params) => {
  const {
    baseUrl,
    redirectUri,
    codeChallenge,
    state,
    clientId,
  } = params;

  const url = new URL('/login', baseUrl);
  url.searchParams.set('client_id', clientId);
  url.searchParams.set('redirect_uri', redirectUri);
  url.searchParams.set('response_type', 'code');
  url.searchParams.set('code_challenge', codeChallenge);
  url.searchParams.set('state', state);
  url.searchParams.set('prompt', 'select_account');

  return url;
};

/**
 * Exchange authorization code for access token
 * @param {{ apiHost: string, code: string, codeVerifier: string, redirectUri: string, clientId: string }} params
 * @returns {Promise<{ accessToken: string }>}
 * @throws {PortalAuthenticationError} with code 'EXCHANGE_FAILED'
 */
async function exchangeCodeForToken(params) {
  const { apiHost, code, codeVerifier, redirectUri, clientId } = params;

  try {
    const { data } = await axios.post(`${apiHost}/portal/auth/token`, {
      grantType: 'authorization_code',
      redirectUri,
      clientId,
      code,
      codeVerifier,
    });

    return data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      // We let the description be empty since there is a mapping function that has a fallback for it
      let description = '';
      if (axios.isAxiosError(error)) {
        if (error.response) {
          description = error.response.data?.message || '';
        } else {
          description = error.message || '';
        }
      } else {
        description = error.message || '';
      }

      throw new PortalAuthenticationError(
        PORTAL_ERROR_CODES.exchangeFailed,
        description,
      );
    }

    throw new PortalAuthenticationError(
      PORTAL_ERROR_CODES.exchangeFailed,
      error.message || '',
    );
  }
}

/**
 * Initialize OAuth client for portal login
 * Returns a client object with methods to get login URL and exchange token
 *
 * @param {{ loginUrl: string, loginApiUrl: string, clientId: string }} options
 * @returns {Promise<{
 *   getLoginUrl: () => string,
 *   obtainTokens: () => Promise<{ accessToken: string }>,
 *   cleanup: () => void
 * }>}
 */
async function initAuthClient(options) {
  const { loginUrl, loginApiUrl, clientId } = options;

  const codeVerifier = generateCodeVerifier();
  const codeChallenge = generateCodeChallenge(codeVerifier);
  const state = generateLoginState();

  // Start loopback server with configuration
  const { port, close: closeLoopbackServer, loginCompletePromise } = await startLoopbackServer({
    address: PORTAL_LOGIN_CONFIGS.loopbackServer.address,
    port: PORTAL_LOGIN_CONFIGS.loopbackServer.autoAssignedPort,
  });

  const redirectUri = `http://${PORTAL_LOGIN_CONFIGS.loopbackServer.address}:${port}/callback`;

  const portalLoginUrl = generateLoginUrl({
    baseUrl: loginUrl,
    redirectUri,
    codeChallenge,
    state,
    clientId,
  });

  const getLoginUrl = () => portalLoginUrl.toString();

  const obtainTokens = async () => {
    let timeoutId;
    let sigintHandler;

    try {
      // Prevent infinite waiting for the browser response
      const timeoutPromise = new Promise((_, reject) => {
        timeoutId = setTimeout(() => {
          reject(
            new PortalAuthenticationError(PORTAL_ERROR_CODES.loginTimeout)
          )
        }, PORTAL_LOGIN_CONFIGS.loginTimeoutMs);
      });

      // Create cancellation promise for user termination (Ctrl+C)
      const cancellationPromise = new Promise((_, reject) => {
        sigintHandler = () => {
          reject(
            new PortalAuthenticationError(PORTAL_ERROR_CODES.loginCancelled)
          );
        };
        process.once('SIGINT', sigintHandler);
      });

      const { code, state: returnedState } = await Promise.race([
        loginCompletePromise,
        timeoutPromise,
        cancellationPromise,
      ]);

      // Validate state to prevent CSRF attacks
      if (state !== returnedState) {
        throw new PortalAuthenticationError(PORTAL_ERROR_CODES.stateMismatch);
      }

      // Exchange authorization code for access token
      const { accessToken } = await exchangeCodeForToken({
        apiHost: loginApiUrl,
        code,
        codeVerifier,
        redirectUri,
        clientId,
      });

      return { accessToken };
    } catch (error) {
      if (error instanceof PortalAuthenticationError) throw error;

      throw new PortalAuthenticationError(
        PORTAL_ERROR_CODES.unknownError,
        error.message || '',
      );
    } finally {
      // Cleanup resources
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
      if (sigintHandler) {
        process.removeListener('SIGINT', sigintHandler);
      }
      closeLoopbackServer();
    }
  };

  return {
    getLoginUrl,
    obtainTokens,
    cleanup: () => { closeLoopbackServer() },
  };
}

module.exports = {
  initAuthClient,
};
