// dbdocs/src/commands/login.js
const { Command, Flags } = require('@oclif/core');
const open = require('open');
const inquirer = require('inquirer').default;
const axios = require('axios');
const ora = require('ora');
const netrc = require('netrc-parser').default;
const { vars } = require('../vars');
const { isValidEmail } = require('../validators/email');
const { isValidOtp } = require('../validators/otp');
const { LOGIN_METHODS, LOGIN_STRATEGIES, PORTAL_ERROR_CODES, PORTAL_LOGIN_CONFIGS } = require('../utils/constants');
const { initAuthClient } = require('../services/auth/portal');
const { sleep } = require('../utils/helper');
const { openUrlInExternalBrowser } = require('../utils/browser');

const TIMEOUT_BEFORE_OPEN_BROWSER_MS = 500;

async function askForOtp (spinner, shortLivedToken, email) {
  const cliSpinner = spinner;
  const otpAnswer = await inquirer.prompt([
    {
      message: 'Please input OTP code sent to the email:',
      name: 'otp',
      validate: (otp) => (isValidOtp(otp) ? true : 'OTP invalid, please try again'),
    },
  ]);

  const { otp } = otpAnswer;

  cliSpinner.text = 'Login to your account';
  cliSpinner.start();
  const { data: { token } } = await axios.post(
    `${vars.apiUrl}/auth`,
    {
      params: {
        authCode: otp,
        provider: 'email',
        shortLivedToken,
        loginEmail: email,
      },
    },
  );
  cliSpinner.succeed();

  return token;
}

async function loginViaEmail (spinner) {
  const cliSpinner = spinner;
  const emailAnswer = await inquirer.prompt([
    {
      message: 'Your email:',
      name: 'email',
      validate: (email) => (isValidEmail(email) ? true : 'Email invalid, please try again'),
    },
  ]);

  const { email } = emailAnswer;

  cliSpinner.text = 'Request email authentication';
  cliSpinner.start();
  const { data: { shortLivedToken } } = await axios.post(
    `${vars.apiUrl}/auth/email`,
    {
      params: {
        email,
      },
    },
  );
  cliSpinner.succeed();

  // TODO: allow user to retry input the OTP
  const token = await askForOtp(spinner, shortLivedToken, email);

  return token;
}

function getPortalErrorMessage(errorCode, errorDescription = '') {
  // If we have errorDescription, use it (this is the main display)
  if (errorDescription) return errorDescription;

  // Otherwise, fallback to default mapped messages
  const errorMessages = {
    [PORTAL_ERROR_CODES.loginTimeout]: 'Login timed out after 5 minutes. Please try again.',
    [PORTAL_ERROR_CODES.loginCancelled]: 'Login cancelled.',
    [PORTAL_ERROR_CODES.stateMismatch]: 'Security validation failed. Please try again.',
    [PORTAL_ERROR_CODES.invalidCallback]: 'Invalid authentication callback received. Please try again.',
    [PORTAL_ERROR_CODES.exchangeFailed]: 'Failed to exchange authorization code. Please try again.',
    [PORTAL_ERROR_CODES.loopbackServerError]: 'Failed to start callback server. Please try again.',
    [PORTAL_ERROR_CODES.unknownError]: 'Login failed. Please try again.',
  };

  return errorMessages[errorCode] || 'An unexpected error occurred. Please try again.';
}

class LoginCommand extends Command {
  /**
   * Detect which login strategy to use
   * Priority: CLI flag > env var > default (portal)
   * @param {boolean} legacyFlag - The --legacy flag value
   * @param {string} loginStrategy - The login strategy from environment variable
   * @returns {'portal' | 'legacy'}
   */
  detectStrategy (legacyFlag, loginStrategy) {
    if (legacyFlag) return LOGIN_STRATEGIES.legacy;

    return loginStrategy;
  }

  /**
   * Execute portal login flow (OAuth 2.0 with PKCE)
   * @param {ora.Ora} spinner - The ora spinner instance
   * @returns {Promise<string>} Access token
   * @throws {Error} If login fails (after displaying error to user)
   */
  async executePortalLogin (spinner) {
    let authClient = null;

    try {
      this.log('Preparing authentication...');

      authClient = await initAuthClient({
        loginUrl: vars.loginUrl,
        loginApiUrl: vars.loginApiUrl,
        clientId: PORTAL_LOGIN_CONFIGS.clientId,
      });

      const loginUrl = authClient.getLoginUrl();

      spinner.text = 'Attempting to open authentication page in your browser...';
      spinner.start();

      try {
        await sleep(TIMEOUT_BEFORE_OPEN_BROWSER_MS);
        await openUrlInExternalBrowser(loginUrl);

        spinner.succeed();
      } catch (err) {
        spinner.warn();
        spinner.warn(`Could not open browser automatically. Please navigate to:\n${loginUrl}`);
      }

      spinner.text = 'Waiting for authentication... (Press CTRL+C to cancel)';
      spinner.start();

      const { accessToken } = await authClient.obtainTokens();

      spinner.text = 'Authentication successful';
      spinner.succeed();

      return accessToken;
    } catch (error) {
      const errorMessage = error.code
        ? getPortalErrorMessage(error.code, error.errorDescription)
        : `Login failed: ${error.message || 'Unknown error'}`;

      this.error(errorMessage);
    } finally {
      // Always cleanup
      if (authClient) {
        authClient.cleanup();
      }
    }
  }

  /**
   * Execute legacy login flow (Email OTP or Social)
   * @param {ora.Ora} spinner - The ora spinner instance
   * @returns {Promise<string>} Access token
   * @throws {Error} If login fails (after displaying error to user)
   */
  async executeLegacyLogin (spinner) {
    try {
      // Show deprecation warning
      this.warn('⚠️ You are using the legacy login. You can remove the --legacy flag for automatic browser authentication.');

      // Prompt for login method
      const loginMethodAnswer = await inquirer.prompt([
        {
          type: 'rawlist',
          message: 'Choose a login method:',
          choices: [
            { name: 'Email', value: LOGIN_METHODS.EMAIL },
            { name: 'GitHub/Google', value: LOGIN_METHODS.GITHUB_GOOGLE },
          ],
          name: 'loginMethod',
        },
      ]);

      const { loginMethod } = loginMethodAnswer;

      let authToken;

      if (loginMethod === LOGIN_METHODS.EMAIL) {
        authToken = await loginViaEmail(spinner);
      } else {
        await open(`${vars.hostUrl}/login/cli`, { wait: false });
        const answer = await inquirer.prompt([
          {
            message: 'Please input your authentication token: ',
            name: 'authToken',
          },
        ]);

        authToken = answer.authToken;
      }

      return authToken;
    } catch (error) {
      // Handle legacy-specific errors
      let errorMessage = error.message || 'Login failed. Please try again.';

      if (error.response) {
        const { error: apiError } = error.response.data;
        switch (apiError.name) {
          case 'OtpInvalidError':
            errorMessage = 'Invalid OTP. Please login again.';
            break;

          case 'OtpUsedError':
            errorMessage = 'OTP used recently. Please login again after couple minutes.';
            break;

          case 'EmailDeliveryFailedError':
            errorMessage = 'Email delivery failed. Please login again.';
            break;

          default:
            errorMessage = error.message || 'Login failed. Please try again.';
            break;
        }
      }

      this.error(errorMessage);
    }
  }

  /**
   * Verify token with API
   * @param {string} token - Access token to verify
   * @returns {Promise<object>} User account object
   * @throws {Error} If token is invalid
   */
  async verifyToken (token) {
    try {
      const { data: { account } } = await axios.get(`${vars.apiUrl}/account`, {
        headers: {
          Authorization: token,
          'Authorization-Method': 'login',
        },
      });

      return account;
    } catch (error) {
      if (error.response) {
        const { error: apiError } = error.response.data;
        switch (apiError.name) {
          case 'TokenExpiredError':
            throw new Error('Your token has expired. Please login again.');

          case 'InvalidAuthToken':
            throw new Error('Invalid token. Please login again.');

          default:
            throw new Error('Failed to verify token. Please try again.');
        }
      }

      throw error;
    }
  }

  /**
   * Store token to ~/.netrc file
   * @param {string} token - Access token to store
   * @param {object} user - User account object with email
   * @returns {Promise<void>}
   */
  async storeToken (token, user) {
    const { apiHost } = vars;
    await netrc.load();

    const previousEntry = netrc.machines[apiHost];
    if (!previousEntry) {
      netrc.machines[apiHost] = {};
    }

    netrc.machines[apiHost].login = user.email;
    netrc.machines[apiHost].password = token;
    await netrc.save();
  }

  async run () {
    const { flags } = await this.parse(LoginCommand);
    const spinner = ora({});

    try {
      // 1. Detect strategy
      const strategy = this.detectStrategy(flags.legacy, vars.loginStrategy);

      // 2. Execute strategy (returns token)
      let accessToken;
      if (strategy === LOGIN_STRATEGIES.legacy) {
        accessToken = await this.executeLegacyLogin(spinner);
      } else {
        accessToken = await this.executePortalLogin(spinner);
      }

      // 3. Verify token
      spinner.text = 'Validate token';
      spinner.start();
      const user = await this.verifyToken(accessToken);
      spinner.succeed();

      // 4. Store token
      spinner.text = 'Save credential';
      spinner.start();
      await this.storeToken(accessToken, user);
      spinner.succeed();

      this.log('\nDone.');
    } catch (error) {
      // Cleanup spinner if still running
      if (spinner.isSpinning) {
        spinner.fail();
      }

      this.error(error.message || 'Something wrong. Please try again.');
    }
  }
}

LoginCommand.description = 'login to dbdocs';

LoginCommand.flags = {
  legacy: Flags.boolean({
    description: 'Use legacy login methods: Email OTP or Google/GitHub with manual token copy',
    default: false,
  }),
};

module.exports = LoginCommand;
