// dbdocs/src/commands/ls.js
const { Command, CliUx } = require('@oclif/core');
const chalk = require('chalk');
const { vars } = require('../vars');
const verifyToken = require('../utils/verifyToken');
const { getProjectsByOrg } = require('../utils/org');
const { getAssociatedOrgs } = require('../utils/org');
const { PROJECT_SHARING_TEXT } = require('../utils/constants');
const { getProjectUrl } = require('../utils/helper');

const ROLE = {
  viewer: 'viewer',
  editor: 'editor',
  admin: 'admin',
};

const ROLE_DESCRIPTIONS = {
  viewer: 'View only',
  editor: 'View and edit',
  admin: 'Administrator',
  noAccess: 'No access',
};

/**
 * The `LsCommand` class is responsible for listing all projects associated with the user.
 * It retrieves projects from various orgs (owned, member, and guest) and displays them in a formatted table,
 * including details about project ownership, user permissions, and general access control.
 */
class LsCommand extends Command {
  /**
   * Displays a formatted table of projects for a given org.
   * It calculates column widths dynamically and presents project details, ownership, permissions, and general access control.
   *
   * @param {Object} org - The org object associated with the projects.
   * @param {Object} user - The current user object.
   * @param {boolean} isOrgOwner - Indicates if the current user is the owner of the org.
   */
  showProjects (org, user, isOrgOwner) {
    const projects = this.flattenAllOrgProjects(org);

    // The starting width for each column (equals to the column names)
    // These widths are then increased to contains the values in each column
    // The columns are
    // 1. `url`
    // 2. `updated at`
    // 3. `owner`
    const defaultColumnsWidth = [3, 12, 5];

    const [maxUrlWidth, maxUpdatedAtWidth, maxOwnerWidth] = projects.reduce((accumulator, project) => {
      const url = getProjectUrl(vars.hostUrl, org.name, project.urlName);
      const updatedAt = (new Date(project.updatedAt)).toLocaleString();
      const owner = this.getProjectOwnerStr(project, user);
      return [
        Math.max(accumulator[0], url.length),
        Math.max(accumulator[1], updatedAt.length),
        Math.max(accumulator[2], owner.length),
      ];
    }, defaultColumnsWidth);

    this.log(chalk.bold(org.name));

    const projectNameMinWidth = 20;
    const projectPermissionMinWidth = 15;
    const projectGeneralAccessMinWidth = 20;

    CliUx.ux.table(projects, {
      name: {
        minWidth: projectNameMinWidth,
      },
      owner: {
        minWidth: maxOwnerWidth + 2,
        get: (project) => this.getProjectOwnerStr(project, user),
      },
      permission: {
        minWidth: projectPermissionMinWidth,
        // The diagram to determine the user permission in a project is in the link below
        // https://www.notion.so/holistics/Roles-Permission-control-127f89dc7e4980fea7efc0526a3d93f2?source=copy_link#13ff89dc7e498033a568f4de6fa4c280
        get: (project) => {
          if (this.isProjectOwner(project, user)) return ROLE_DESCRIPTIONS.admin;

          const isInvitedAsEditor = project.role?.name === ROLE.editor;

          // If the user is a guest (invited to the project, but not invited to the org)
          // then the final permission is the invited permission
          const isGuest = !org.orgRoleName && !isOrgOwner;
          if (isGuest) return isInvitedAsEditor ? ROLE_DESCRIPTIONS.editor : ROLE_DESCRIPTIONS.viewer;

          // Org viewers never have permission to edit project
          const isOrgEditor = isOrgOwner || (org.orgRoleName === ROLE.editor);
          if (!isOrgEditor) return ROLE_DESCRIPTIONS.viewer;

          const isProjectSharedForEditInOrg = project.OrgRole?.name === ROLE.editor;

          return (isProjectSharedForEditInOrg || isInvitedAsEditor) ? ROLE_DESCRIPTIONS.editor : ROLE_DESCRIPTIONS.viewer;
        },
      },
      generalAccess: {
        header: 'Access Control',
        minWidth: projectGeneralAccessMinWidth,
        get: (project) => {
          const canViewGeneralAccess = this.isProjectOwner(project, user) && project.generalAccessType;
          return canViewGeneralAccess
            ? PROJECT_SHARING_TEXT[project.generalAccessType]
            : 'N/A';
        },
      },
      url: {
        minWidth: maxUrlWidth + 2,
        get: (project) => chalk.cyan(getProjectUrl(vars.hostUrl, org.name, project.urlName)),
      },
      updatedAt: {
        minWidth: maxUpdatedAtWidth + 2,
        header: 'Last updated',
        get: (project) => (new Date(project.updatedAt)).toLocaleString(),
      },
    }, {
      printLine: this.log.bind(this),
    });

    if (!projects.length) CliUx.ux.log('', '(empty)');
    this.log();
  }

  // eslint-disable-next-line class-methods-use-this
  isProjectOwner (project, user) {
    return user.id && project.Owner?.id === user.id;
  }

  getProjectOwnerStr (project, user) {
    return this.isProjectOwner(project, user)
      ? `${user.email} (you)`
      : project.Owner?.email || '';
  }

  /**
   * Consolidates all types of projects (owned, shared, and private) associated with a given org into a single list.
   *
   * @param {Object} org - The org object, potentially containing `projects`, `sharedProjects`, and `privateProjects` arrays.
   * @returns {Array<Object>} An array containing all projects (owned, shared, and private) from the org.
   */
  // eslint-disable-next-line class-methods-use-this
  flattenAllOrgProjects (org) {
    const { projects = [], sharedProjects = [], privateProjects = [] } = org;
    return [...projects, ...sharedProjects, ...privateProjects];
  }

  /**
   * Main entry point for the `ls` command.
   * It retrieves and displays all projects associated with the user across various orgs (owned, member, and guest).
   *
   * @returns {Promise<void>}
   */
  async run () {
    try {
      const { authConfig, account } = await verifyToken();
      const { ownedOrgs, memberOrgs, guestOrgs } = await getAssociatedOrgs(authConfig);

      const getOwnedOrgsProjectsPromise = Promise.all(ownedOrgs.map((org) => getProjectsByOrg(org.name, authConfig)));

      const getJoinedOrgsPromise = Promise.all(memberOrgs.map(async ({ Org, Role }) => {
        const orgWithProjects = await getProjectsByOrg(Org.name, authConfig);
        return { ...orgWithProjects, orgRoleName: Role.name };
      }));

      const getGuestOrgsProjectsPromise = Promise.all(guestOrgs.map((org) => getProjectsByOrg(org.name, authConfig)));

      const [
        ownedOrgsProjects,
        joinedOrgs,
        guestOrgsProjects,
      ] = await Promise.all([getOwnedOrgsProjectsPromise, getJoinedOrgsPromise, getGuestOrgsProjectsPromise]);

      ownedOrgsProjects.forEach((org) => this.showProjects(org, account, true));
      joinedOrgs.forEach((org) => this.showProjects(org, account, false));
      guestOrgsProjects.forEach((org) => this.showProjects(org, account, false));
    } catch (err) {
      this.error(err);
    }
  }
}

LsCommand.description = 'list projects';

module.exports = LsCommand;
